### ISSUE TITLE
Problems running Dafny on Apple M1 devices (`libhostpolicy.dylib`, "no suitable image found")

### ISSUE THREAD

**Issue by cpitclaudel** [dafny-lang/dafny#1941]
Moving the discussion here from https://github.com/dafny-lang/ide-vscode/issues/146.

Some users are reporting issues with Running Dafny's releases on Apple M1 devices.  The error looks like this:

```
Failed to load /Users/<user>/.vscode/extensions/dafny-lang.ide-vscode-2.3.0/out/resources/3.5.0/dafny/libhostpolicy.dylib, error: dlopen(/Users/<user>/.vscode/extensions/dafny-lang.ide-vscode-2.3.0/out/resources/3.5.0/dafny/libhostpolicy.dylib, 1): no suitable image found.  Did find:
    /Users/<user>/.vscode/extensions/dafny-lang.ide-vscode-2.3.0/out/resources/3.5.0/dafny/libhostpolicy.dylib: mach-o, but wrong architecture
    /Users/<user>/.vscode/extensions/dafny-lang.ide-vscode-2.3.0/out/resources/3.5.0/dafny/libhostpolicy.dylib: mach-o, but wrong architecture
An error occurred while loading required library libhostpolicy.dylib from [/Users/<user>/.vscode/extensions/dafny-lang.ide-vscode-2.3.0/out/resources/3.5.0/dafny/]
[Info  - 9:57:07 AM] Connection to server got closed. Server will restart.
(... 5 times)
[Error - 9:57:07 AM] Connection to server got closed. Server will not be restarted.
```

For now, the best-known workaround is to build from source.

## Current status

1. We are investigating the issue.  It seems that installing Dafny with Brew works, but this is not a complete workaround because the brew package does not include Dafny's LSP server.  Brew installs the exact same binaries as our releases, so it has to come down to some configuration issue.
2. We (at least @keyboardDrummer) have a repro on one of our machines.
3. Building from source is reported to work.

## Workaround

If you run into this issue, we recommend that you [build Dafny from source](https://github.com/dafny-lang/dafny/wiki/INSTALL#mac-source).  You can then configure VSCode or Emacs to use the newly built binary.

**Comment by MikaelMayer** [comment]
With @prvshah51 , we just implemented and successfully tested an automated workaround to this problem here: https://github.com/dafny-lang/ide-vscode/pull/155
Basically, if it detects the unsupported architecture, it will download Dafny, build it, and replace the files by those freshly built.
It works for simple files, all the pipeline works, but for some more complex files we still found some issues, but they are so far related to Non-Windows + typing slowly.

**Comment by atomb** [comment]
We're now building Dafny for arm64, included with 3.10.0 and recent nightly builds. There's an IDE plugin PR (dafny-lang/ide-vscode#316) that'll install those, and should allow us to close this issue.
